### DISCUSSION TITLE
Question: Ping ESP32C3 is bad after hard reboot, but good after soft reboot

### DISCUSSION THREAD

**Discussion by TRadigk** [micropython/micropython#12462] [ESP32] [answered]
Hi guys,

I hope you have one or maybe more ideas for my issue with my ESP32C3 project.
It is a simple project that connects the board to my WIFI and starts a microdot server on which I have registered some routes. The server is started and as microdot does, it runs forever. Fairly simple.
But while testing I got barely an answer back, testing ping on the ip address showed either "timeout" or values above 2000 ms. But when I open Thonny, press CTRL+C to stop and CTRL+D to restart and wait for the reconnection to be done pings are in range of 5 to 53  ms range.

I've experimented with an automated reboot after a hard reset with sys.exit(), without success. The esp is running micropython 1.20-379-ga18d62e06, a nightly built and flashed some weeks ago.

This is basically the code I am executing (stripped down to the very basics, I have omitted content of variables):

```python
import network  # type:ignore
import Microdot

if __name__ == "__main__":
    wlan = network.WLAN(network.STA_IF)
    wlan.active(True)
    wlan.disconnect()
    wlan.config(hostname="esp32")
    wlan.connect(str.encode("x"), str.encode("y"))

    while not wlan.isconnected():
        sleep(1)
    app = Microdot()
    app.run(host="address is set", port="port is set", debug=True)
```

There is some async code involved during bootup, which is not shown here, but I have tried to do everything synchronously without any other effect than the already described one.

I have the suspicion the WIFI module is involved somehow, because the ping is still working after pressing CTRL+C 😮 while being connected with Thonny.

The ping response is important for my client application to know that in general the server is kind of "available".

**Comment by davefes** [comment]
Maybe, try:
```py
while not wlan.isconnected():
    sleep(1)
    print (f'network config: {wlan.ifconfig()}')
```

**Reply by TRadigk** [reply]
Also is part of the code I omitted. wlan.ifconfig() is called.

**Reply by davefes** [reply]
So ifconfig() is the same in both cases?

**Reply by TRadigk** [reply]
No, the answers are of course not the same, since I only can make use of it, when the connection is established.
Until then the values are 
`('0.0.0.0', '0.0.0.0', '0.0.0.0', '192.168.0.1')` 
Which is expected, since the first item of the tuple is the ip address the controller not yet has gotten.

Please see the answer from shariltumin below, he had the right point: set the power mode to "WLAN.PM_NONE" to achieve reliable results.

**Comment by shariltumin** [comment, marked as answer]
What is the power management mode of your STA?

You can try turning it off completely and see how it works for you.

```
MicroPython v1.20.0-332-gb714f4181-dirty on 2023-08-20; ESP32C3 LUATOS-A12-USB (KAKI5) with ESP32C3
>>> 
>>> import network
>>> sta = network.WLAN(network.STA_IF)
>>> sta.active()
False
>>> sta.active(True)
True
>>> sta.config('pm')
2
>>> sta.config(pm=sta.PM_NONE)
>>> sta.config('pm')
0
```

**Reply by TRadigk** [reply]
hi @shariltumin  

yes, setting the value "0" actively leads to reliable ping results. I did take the time to check the PM value in both scenarios (Soft rebooted, reset via hardware) but could not find any difference. It was always "1".
My issue is kind of resolved, since the project can continue, but I still wonder if this is an issue worth creating. To me it looks like the power mode is set differently after hard reset - my code never configured it, but that is also true for soft reboot. What do you think?

**Reply by shariltumin** [reply]
PM will retain its value after a soft rest, but will be reset to its default value (1) after a hard rest.

**Reply by TRadigk** [reply]
But why, if not set, do I get different  (ping test) results on soft reboot and hard reset? And what is reflected by config("pm") leading to `PM_PERFORMANCE`, though the ping result tells me, after soft reboot, it is rather `PM_NONE` than performance? Is the method not actually reading hardware values? Is the pm value set somewhere hidden in other methods?

**Comment by bixb922** [comment]
Soft reset does not reset the wifi subsystem, hard reset does. One advantage is that after a soft reset, connecting is usually much faster.

To reset the wifi subsystem try ```wlan.active(False) ``` just before setting ```wlan.active(True)```. That usually returns the WLAN to a initial state.

**Comment by davefes** [comment]
Doing this at the beginning of my connect sequence seems to be more reliable:
```py
if sta_if.isconnected():
    sta_if.disconnect()
    print ('started in the connected state, but now disconnected')
else:
   print ('started in the disconnected state')
```
Or try glenn20's wifi.py
https://github.com/glenn20/micropython-espnow-utils/blob/main/src/wifi.py
I added this in his disconnect block
```py
def disconnect():
    _sta.disconnect()
    wait_for(lambda: not _sta.isconnected(), 5)
   _sta.active(False)  # added
```
